### PR TITLE
SQL Server LIKE filter escaping #143

### PR DESCRIPTION
**Pull request by stmeyer**
Changes to search in SQL Server with better escaping. I check only the end of the class name, so that everone is able to derive a own implementation of the base class.

### CODE REVIEW

**Comment by dereuromark** [comment]
Looks good to me.

**Comment by ADmad** [comment]
I don't really like how the driver type guessing will be done on every call to `_formatWildcards()`. 

We can add a `driver` config to the filter. The driver guessing code can be moved into a separate method and called just from `process()` and set the `driver` config if not already set.

**Comment by dereuromark** [comment]
Thats a valid concern. Sounds like a better approach.

File: src/Model/Filter/Like.php [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
+use InvalidArgumentException;
 use Search\Manager;
 
 class Like extends Base
 {
 
     /**
+     * driver to do escaping
+     *
+     * @var type
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: FQCN to type missing

File: src/Model/Filter/Like.php [hunk 2]
```diff
@@ -11,7 +11,7 @@ class Like extends Base
     /**
      * driver to do escaping
      *
-     * @var type
+     * @var Search\Model\Filter\EscapeDriver\Base
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: FQCN starts with \

File: src/Model/Filter/EscapeDriver/Normal.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?php
+namespace Search\Model\Filter\EscapeDriver;
+
+class Normal extends Base
```
Comment (dereuromark) [inline src/Model/Filter/EscapeDriver/Normal.php]: Is "Default" a better name maybe?

**Comment by dereuromark** [comment]
Personally, I dont like those short suffix-less class names (class Default extends Base).
I rather prefer sth like \Model\Filter\Escaper\ and
- BaseEscaper
- DefaultEscaper
- SqlserverEscaper

etc.
Those are more speaking and if search for easier to find between tons of "Default" or "Base" classes.

PS: I know that all the filters have the same issue and are missing the suffix. So for the sake of consistency I am fine with the current code - just saying I would have named it a bit differently.

**Comment by stmeyer** [comment]
I added the suffix because I prefer it, too.

If this is ok now I will add some documentation for the escapers, because everyone is able to write on escapers if needed.

Additionally should I add a "NothingEscaper" that leaves the search string as is?

**Comment by dereuromark** [comment]
Isnt nothing already possible by declaring the Wildcards to false? I would Not add it then twice.

File: src/Model/Filter/Like.php [hunk 4]
```diff
@@ -1,12 +1,22 @@
...
 
 class Like extends Base
 {
 
     /**
+     * driver to do escaping
+     *
+     * @var \Search\Model\Filter\EscapeDriver\Base
+     */
+    protected $escapeDriver;
```
Comment (ADmad) [inline src/Model/Filter/Like.php]: Please change to `$_escapeDriver` for consistency.

**Comment by stmeyer** [comment]
Any more hints or ready to merge?

**Comment by ADmad** [comment]
There's lot of code duplication between `DefaultEscaper::formatWildCards()` and `SqlserverEscaper::formatWildCards()`. Would be nice if that could be improved.

Also I think we can do away with the `BaseEscaper` and just have `DefaultEscaper` and `SqlserverEscaper` with `SqlserverEscaper extends DefaultEscaper`.

File: src/Model/Filter/Like.php [hunk 5]
```diff
@@ -143,26 +155,34 @@ protected function _wildcards($value)
...
         }
-        if ($from) {
-            // Escape first
-            $value = str_replace($from, $to, $value);
-            // Replace wildcards
-            $value = str_replace($substFrom, $substTo, $value);
+
+        $class = $this->config('escapeDriver');
+        $className = App::className($class, 'Model/Filter/EscapeDriver', 'Escaper');
+        if (!class_exists($className)) {
```
Comment (ADmad) [inline src/Model/Filter/Like.php]: `App::className` already does a `class_exists()` check and returns `false` on failure, so this should be just `if (!$className) {`

**Comment by stmeyer** [comment]
last travis build did not update the status. are you able to update it?

File: src/Model/Filter/Like.php [hunk 6]
```diff
@@ -1,12 +1,22 @@
...
+use InvalidArgumentException;
 use Search\Manager;
 
 class Like extends Base
 {
 
     /**
+     * driver to do escaping
+     *
+     * @var \Search\Model\Filter\EscapeDriver\Base
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: Base? Maybe we should have an interface then instead?

File: src/Model/Filter/Like.php [hunk 7]
```diff
@@ -143,26 +155,34 @@ protected function _wildcards($value)
...
+    }
+
+    /**
+     * set configuration for escape driver name
+     *
+     * @return void
+     */
+    protected function _setEscapeDriver()
+    {
+        if ($this->config('escapeDriver') === null) {
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: Should this be "escaper" for consistency with the class name? Do we need the word "driver" in here? //cc @ADmad

Comment (dereuromark) [inline src/Model/Filter/Like.php]: Also the folder name doesnt have to be EscapeDriver but Escaper then if we change this.

File: src/Model/Filter/Like.php (hunk 6 above)
Comment (dereuromark) [inline src/Model/Filter/Like.php]: formatWildcards() is the only method as per contract then.

Comment (stmeyer) [inline src/Model/Filter/Like.php]: Yes, interface is usefull

**Comment by dereuromark** [comment]
Nice work!
Are you able to squash the commits?
I think we can otherwise also do this on merge with github.

**Comment by stmeyer** [comment]
It would be nice if you can do it 😄

File: src/Model/Filter/Like.php [hunk 8]
```diff
@@ -1,12 +1,22 @@
...
 
+use Cake\Core\App;
+use InvalidArgumentException;
 use Search\Manager;
 
 class Like extends Base
 {
 
     /**
+     * driver to do escaping
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: Escaper to be used.

File: src/Model/Filter/Like.php [hunk 9]
```diff
@@ -51,6 +62,7 @@ public function process()
             return;
         }
 
+        $this->_setEscapeDriver();
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: Should this be $this->_setEscaper()?

**Comment by dereuromark** [comment]
@ADmad This is good to go, right?

**Comment by stmeyer** [comment]
@ADmad any news?

**Comment by ADmad** [comment]
Sorry forgot about this. Will review again soon and merge.

File: src/Model/Filter/Like.php [hunk 10]
```diff
@@ -1,12 +1,22 @@
...
 {
 
     /**
+     * Escaper to be used.
+     *
+     * @var \Search\Model\Filter\Escaper\EscaperInterface
+     */
+    protected $_escaper;
+
+
```
Comment (ADmad) [inline src/Model/Filter/Like.php]: Extra empty line

**Pull request merged**